# Analysis of Bestsellers
󠀰Virus corona yang kehadirannya mengejutkan seluruh dunia telah mengubah rutinitas harian semua orang.󠀲󠀡󠀠󠀦󠀥󠀨󠀧󠀢󠀳󠀰 Kini, warga kota tidak lagi menghabiskan waktu luangnya di luar rumah seperti pergi ke kafe dan mal. Mereka lebih sering berada di rumah, menghabiskan waktunya dengan membaca buku.󠀲󠀡󠀠󠀦󠀥󠀨󠀧󠀣󠀳󠀰 

Proyek kali ini diberikan sebuah database dari salah satu perusahaan startup untuk mengembangkan aplkikasi baru bagi para pecinta buku.󠀲󠀡󠀠󠀦󠀥󠀨󠀧󠀥󠀳󠀰 Database tersebut berisi data tentang buku, penerbit, penulis, serta rating dan ulasan pelanggan atas buku terkait.󠀲󠀡󠀠󠀦󠀥󠀨󠀧󠀦󠀳󠀰 Informasi ini akan digunakan dalam membuat penawaran harga untuk sebuah produk baru. 


## Tugas
Berikut beberapa tugasnya yaitu:
- Menghitung jumlah buku yang dirilis setelah tanggal 1 Januari 2000.󠀲󠀡󠀠󠀦󠀥󠀩󠀠󠀩󠀳
- Menghitung jumlah ulasan pengguna dan rating rata-rata untuk setiap buku.󠀲󠀡󠀠󠀦󠀥󠀩󠀡󠀠󠀳
- Mengidentifikasi penerbit yang telah menerbitkan jumlah buku terbanyak dengan jumlah halaman lebih dari 50.
- Mengidentifikasi penulis dengan rating rata-rata buku tertinggi: temukan buku dengan rating minimal 50.󠀲󠀡󠀠󠀦󠀥󠀩󠀡󠀢󠀳
- Menghitung jumlah rata-rata teks ulasan di antara pengguna yang memberi rating terhadap lebih dari 50 buku.

## Deskripsi data
`books` 
󠀰Berisi data tentang buku:
- book_id — ID buku
- author_id — ID penulis
- title — judul buku
- num_pages — jumlah halaman
- publication_date — tanggal penerbitan
- publisher_id — ID penerbit

`authors`
Berisi data tentang penulis:
- author_id — ID penulis
- author — nama penulis

`publishers`
Berisi data tentang penerbit:
- publisher_id — ID penerbit
- publisher — nama penerbit

`ratings`
󠀰Berisi data tentang rating pengguna:
- rating_id — ID rating
- book_id — ID buku
- username — nama pengguna yang memberi rating buku
- rating

`reviews`
Berisi data tentang ulasan pelanggan:
- review_id — ID ulasan
- book_id — ID buku
- username — nama pengguna yang mengulas buku
- text — teks ulasan

## Persiapan Analisis

In [1]:
# impor library yang diperlukan
import pandas as pd
from sqlalchemy import create_engine

In [2]:
#!pip3 install python-dotenv
#!pip3 install sqlalchemy
#!pip3 install psycopg2-binary

# membuat skrip untuk mengakses database
db_config = {'user': 'practicum_student', # nama pengguna
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # kata sandi
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # port koneksi
             'db': 'data-analyst-final-project-db'}          # nama database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Eksplorasi Data

In [3]:
# melihat 5 baris pertama tabel books
query = '''
         SELECT *
         FROM books
         LIMIT 5
        '''

data = pd.io.sql.read_sql(query, con = engine)
data

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
# melihat 5 baris pertama tabel authors
query = '''
         SELECT *
         FROM authors
         LIMIT 5
        '''

data = pd.io.sql.read_sql(query, con = engine)
data

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
# melihat 5 baris pertama tabel publishers
query = '''
         SELECT *
         FROM publishers
         LIMIT 5
        '''

data = pd.io.sql.read_sql(query, con = engine)
data

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [6]:
# melihat 5 baris pertama tabel ratings
query = '''
         SELECT *
         FROM ratings
         LIMIT 5
        '''

data = pd.io.sql.read_sql(query, con = engine)
data

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
# melihat 5 baris pertama tabel reviews
query = '''
         SELECT *
         FROM reviews
         LIMIT 5
        '''

data = pd.io.sql.read_sql(query, con = engine)
data

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Observasi Data
- Jumlah buku yang dirilis setelah tanggal 1 Januari 2000.󠀲󠀡󠀠󠀦󠀥󠀩󠀠󠀩󠀳
- Jumlah ulasan pengguna dan rating rata-rata untuk setiap buku.󠀲󠀡󠀠󠀦󠀥󠀩󠀡󠀠󠀳
- Penerbit yang telah menerbitkan jumlah buku terbanyak dengan jumlah halaman lebih dari 50 (hal ini akan membantumu dalam mengecualikan brosur dan publikasi serupa dari analisismu).󠀲󠀡󠀠󠀦󠀥󠀩󠀡󠀡󠀳
- Penulis dengan rating rata-rata buku tertinggi: temukan buku dengan rating minimal 50.󠀲󠀡󠀠󠀦󠀥󠀩󠀡󠀢󠀳
- Jumlah rata-rata teks ulasan di antara pengguna yang memberi rating terhadap lebih dari 50 buku.

### Jumlah buku yang dirilis setelah tanggal 1 Januari 2000

In [8]:
# menghitung jumlah buku
query = '''
         SELECT COUNT(*) AS book_count
         FROM books
         WHERE publication_date > '01-01-2000'
        '''

data = pd.io.sql.read_sql(query, con = engine)
data

book_count
0         819

Jumlah buku yang dirilis setelah tanggal 1 Januari 2000 sebanyak 819 buku.

### Jumlah ulasan pengguna dan rating rata-rata untuk setiap buku

In [9]:
# menghitung jumlah ulasan pengguna dan rating rata-rata untuk setiap buku
query = '''
        SELECT 
            b.title, 
            COUNT(r.rating_id) AS total_reviews, 
            AVG(r.rating) AS avg_rating
        FROM books b
        LEFT JOIN 
            ratings r ON b.book_id = r.book_id
        GROUP BY 
            b.book_id, 
            b.title
        ORDER BY 
            total_reviews DESC;
        '''

data = pd.io.sql.read_sql(query, con = engine)
data

title  total_reviews  \
0                              Twilight (Twilight  #1)            160   
1                  The Hobbit  or There and Back Again             88   
2                               The Catcher in the Rye             86   
3                 Angels & Demons (Robert Langdon  #1)             84   
4    Harry Potter and the Prisoner of Azkaban (Harr...             82   
..                                                 ...            ...   
995                  Naked Empire (Sword of Truth  #8)              2   
996          A Woman of Substance (Emma Harte Saga #1)              2   
997          The Body in the Library (Miss Marple  #3)              2   
998  The Magicians' Guild (Black Magician Trilogy  #1)              2   
999  Disney's Beauty and the Beast (A Little Golden...              1   

     avg_rating  
0      3.662500  
1      4.125000  
2      3.825581  
3      3.678571  
4      4.414634  
..          ...  
995    3.500000  
996    5.000000  
997    4.500000  
998    3.500000  
999    4.000000  

[1000 rows x 3 columns]

Dari tabel di atas dapat diketahui bahwa jumlah ulasan pengguna tertinggi yaitu buku Twilight (Twilight#1) dengan total jumlah ulasan sebanyak 160. Sedangkan jumlah ulasan terendah yaitu buku Disney's Beauty and the Beast yaitu hanya 1 pengguna yang mengulas buku tersebut.

### Penerbit yang telah menerbitkan jumlah buku terbanyak dengan jumlah halaman lebih dari 50

In [10]:
# melihat penerbit yang telah menerbitkan jumlah buku terbanyak dengan jumlah halaman lebih dari 50
query = '''
        SELECT 
            p.publisher,
            COUNT(b.book_id) AS books
        FROM publishers p
        JOIN
            books b ON p.publisher_id = b.publisher_id
        WHERE
            b.num_pages > 50
        GROUP BY
            p.publisher
        ORDER BY
            books DESC
        LIMIT 5;    
        '''

data = pd.io.sql.read_sql(query, con = engine)
data

publisher  books
0             Penguin Books     42
1                   Vintage     31
2  Grand Central Publishing     25
3          Penguin Classics     24
4                    Bantam     19

Tabel di atas memuat 5 penerbit yang telah menerbitkan buku terbanyak dengan jumlah halaman lebih dari 50 lembar. Dapat diketahui bahwa penerbit `Penguin Books` menjadi penerbit yang paling banyak menerbitkan buku.

### Penulis dengan rating rata-rata buku tertinggi, dengan mempertimbangkan buku yang memiliki rating minimal 50

In [11]:
# melihat penulis dengan rating rata-rata buku tertinggi, dengan mempertimbangkan buku yang memiliki rating minimal 50
query = '''
        SELECT 
            a.author, 
            AVG (avg_rating) AS avg_rating
        FROM authors a
        JOIN 
            books b ON a.author_id = b.author_id
        JOIN
        (
            SELECT 
                bq.book_id,
                bq.title,
                COUNT (rq.rating_id) AS rating_count,
                AVG (rq.rating) AS avg_rating
            FROM books bq
            JOIN
                ratings rq ON rq.book_id = bq.book_id
            GROUP BY
                bq.book_id,
                bq.title
                
        ) AS subq ON subq.book_id = b.book_id
        WHERE 
            rating_count >= 50
        GROUP BY 
            a.author_id, 
            a.author
        ORDER BY 
            avg_rating DESC
        LIMIT 5;
        '''

data = pd.io.sql.read_sql(query, con = engine)
data

author  avg_rating
0         J.K. Rowling/Mary GrandPré    4.283844
1  Markus Zusak/Cao Xuân Việt Khương    4.264151
2                     J.R.R. Tolkien    4.258446
3                  Louisa May Alcott    4.192308
4                       Rick Riordan    4.080645

Tabel di atas memuat 5 penulis dengan rating rata-rata tertinggi dengan kriteria jumlah minimal ulasan sebanyak 50. Penulis `J.K. Rowling/Mary GrandPré` menjadi penulis dengan rating rata-rata tertinggi dengan rata-rata 4.28.

### Jumlah rata-rata teks ulasan di antara pengguna yang memberi rating terhadap lebih dari 50 buku

In [12]:
# menghitung jumlah rata-rata teks ulasan di antara pengguna yang memberi rating terhadap lebih dari 50 buku
query = '''
        SELECT 
            AVG (review_count) AS avg_review_lenght
        FROM 
        (
            SELECT 
                r.username,
                COUNT (DISTINCT re.review_id) AS review_count
            FROM
                ratings r
            LEFT JOIN
                reviews re ON r.username = re.username
            WHERE
                r.username IS NOT NULL AND r.book_id IS NOT NULL
            GROUP BY
                r.username
            HAVING COUNT(DISTINCT r.book_id) > 50
        ) AS user_review
        
        '''

data = pd.io.sql.read_sql(query, con = engine)
data

avg_review_lenght
0          24.333333

Rata-rata jumlah teks ulasan di antara pengguna yang telah mengulas lebih dari 50 buku adalah 24.

## Kesimpulan
- Terdapat 819 buku yang dirilis setelah tanggal 1 Januari 2000.
- Buku `Twilight (Twilight#1)` memiliki jumlah ulasan pengguna tertinggi, yaitu 160 ulasan, menunjukkan popularitas dan keterlibatan pembaca yang tinggi.
- Terdapat variasi jumlah ulasan, dari yang tinggi hingga yang rendah, menandakan variasi tingkat popularitas antar buku.
- Penerbit `Penguin Books` merupakan penerbit terkemuka yang paling banyak menerbitkan buku dengan jumlah halaman lebih dari 50. Hal ini dapat menjadi informasi penting dalam analisis bisnis dan strategi pemasaran.
- Penulis `J.K. Rowling/Mary GrandPré` memiliki rating rata-rata tertinggi (4.28) dari buku-buku yang memiliki minimal 50 ulasan. Ini menunjukkan kualitas konsisten dari karya penulis tersebut.
- Rata-rata panjang teks ulasan di antara pengguna yang memberi rating terhadap lebih dari 50 buku adalah 24 karakter. Informasi ini dapat membantu dalam memahami sejauh mana ulasan pengguna secara rinci dan informatif.

## Rekomendasi Bisnis
- Meningkatkan kampanye pemasaran untuk buku yang sudah populer dengan strategi promosi seperti program loyalitas & diskon khusus untuk buku yang memiliki banyak ulasan guna meningkatkan engagement, bundling produk misalnya, menjual seri Twilight dalam satu paket dengan merchandise eksklusif, dan pemasaran berbasis komunitas, seperti mengadakan diskusi buku, fan meet-up, atau event bertema.
- Memanfaatkan tren ulasan pembaca sebagai alat pemasaran dengan menampilkan kutipan ulasan terbaik dalam kampanye iklan.
- Menonjolkan buku-buku dengan sedikit ulasan tetapi memiliki potensi tinggi dengan meningkatkan visibilitas mereka melalui rekomendasi yang dipersonalisasi.
- Memanfaatkan sistem rekomendasi berbasis AI di platform e-commerce untuk mengarahkan pembaca ke buku yang relevan dengan preferensi mereka.
- Kolaborasi dengan influencer atau booktubers/bookstagrammers untuk meningkatkan eksposur buku dengan ulasan rendah tetapi berkualitas tinggi.
- Membangun kemitraan strategis dengan Penguin Books untuk eksklusivitas buku-buku tertentu atau edisi khusus.
- Meningkatkan katalog digital dan fisik dari penerbit ini, dengan kategori khusus di platform penjualan online untuk menarik lebih banyak pelanggan.
- Menyediakan layanan premium atau berlangganan untuk akses eksklusif ke buku-buku penerbit terkemuka dengan harga khusus.
- Menawarkan edisi kolektor atau edisi terbatas dari buku-buku dengan rating tinggi, misalnya buku dengan ilustrasi khusus atau versi hardcover premium.
- Mempromosikan ulang buku-buku dari penulis dengan rating tertinggi seperti embuat program rekomendasi berbasis kualitas, seperti "Top Rated Books of the Year" atau "Highly Reviewed by Readers".
- Mendorong ulasan yang lebih mendalam dengan insentif seperti poin reward atau badge bagi pengguna yang menulis ulasan lebih panjang dan informatif.
- Meningkatkan fitur interaksi di platform buku online, seperti voting ulasan terbaik, diskusi komunitas, atau sistem penghargaan bagi pengguna yang aktif memberikan review.